internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,195 @@
+	errNotFound       = errors.New("metric not found")
+	errPrometheusLint = errors.New("prometheus GatherAndLint")
+	errDiffLabels     = errors.New("labels are incorrect")
+)
+
+// ValidateSpec validates that the Prometheus metrics being exposed from baseplate.go
+// conform to the baseplate spec.
+// metricPrefix should be either "grpc", "http", or "thirft".
+// clientOrServer should be either "client" or "server".
+func ValidateSpec(tb testing.TB, metricPrefix, clientOrServer string) {
+	tb.Helper()
+
+	var batch errorsbp.Batch
+	batch.AddPrefix("validate spec", validateSpec(metricPrefix, clientOrServer))
+
+	if err := batch.Compile(); err != nil {
+		tb.Error(err)
+	}
+}
+
+func validateSpec(prefix, clientOrServer string) error {
+	var batch errorsbp.Batch
+
+	metricFam, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		batch.Add(err)
+		return batch.Compile()
+	}
+
+	metricsUnderTest := buildMetricNames(prefix, clientOrServer)
+
+	for _, m := range metricFam {
+		name := *m.Name
+
+		if _, ok := metricsUnderTest[name]; !ok {
+			continue
+		}
+
+		batch.Add(lintMetric(name))
+		batch.Add(validateName(name, prefix, clientOrServer))
+
+		labels := map[string]struct{}{}
+		for _, metric := range m.GetMetric() {
+			for _, label := range metric.Label {
+				labels[*label.Name] = struct{}{}
+			}
+		}
+		batch.Add(validateLabels(name, prefix, clientOrServer, labels))
+
+		// after we validate the metric, delete it from the set of all
+		// expected metrics so that we can track if any were not found.
+		delete(metricsUnderTest, name)
+	}
+
+	if len(metricsUnderTest) > 0 {
+		batch.Add(fmt.Errorf("%w: %v", errNotFound, keysFrom(metricsUnderTest)))
+	}
+
+	return batch.Compile()
+}
+
+func keysFrom(metrics map[string]struct{}) []string {
+	missing := []string{}
+	if len(metrics) > 0 {
+		for k := range metrics {
+			missing = append(missing, k)
+		}
+	}
+	return missing
+}
+
+// buildMetricNames creates a set of all the expected metrics names for the given prefix
+// that should be registered with Prometheus as defined in the baseplate spec.
+// The prefix will be either thrift, http, or grpc.
+func buildMetricNames(prefix, clientOrServer string) map[string]struct{} {
+	suffixes := []string{"latency_seconds", "requests_total", "active_requests"}
+	names := map[string]struct{}{}
+	for _, suffix := range suffixes {
+		names[prometheus.BuildFQName(prefix, clientOrServer, suffix)] = struct{}{}
+	}
+	return names
+}
+
+// validateName checks the following:
+//   1) the metric name has at least 3 parts separated by "_".
+//     The parts are <namespace>_<client/server>_<name>_<suffix>, where suffix is optional.
+//     Ref: https://prometheus.io/docs/practices/naming
+//   2) the metric name has the correct prefix.
+//   3) the metric contains either "client" or "server" as the second part.
+func validateName(name, prefix, clientOrServer string) error {
+	var batch errorsbp.Batch
+
+	const (
+		separator = "_"
+		partCount = 3
```

Comment:
🔕/super nit: this is the same value in `errLength`. I'd prefer move this const to global and change `errLength` into: `errLength = fmt.Errorf("metric name should have a minimum of %d parts", partCount)`.

also grammar question/nit: should this be `partsCount` instead of `partCount`?